docs/plugins/best_practices.md

```diff
@@ -0,0 +1,175 @@
+    of your dependencies.  Just know that it may not work with the built-in plugin installer. 
+    ````
+
+- **Don't import heavy dependencies at the top of your module.**
+
+    ````{note}
+    This point will be less relevant when we move to the second generation
+    [manifest-based plugin
+    declaration](https://github.com/napari/napari/issues/3115), but it's still a 
+    good idea to delay importing your plugin-specific dependencies and modules until *after* your hookspec has been called.  This helps napari stay quick and responsive
+    at startup.  
+    ````
+
+    Consider the following example plugin:
+
+    ```ini
+    [options.entry_points]
+    napari.plugion =
+      plugin-name = mypackage.napari_plugin
+    ```
+
+    In this example, `my_heavy_dependency_like_tensorflow` will be imported
+    *immediately* when napari is launched, and we search the entry_point
+    `mypackage.napari_plugin` for decorated hook specifications.
+
+    ```py
+    # mypackage/napari_plugin.py
+    from napari_plugin_engine import napari_hook_specification
+    from qtpy.QtWidgets import QWidget
+    from my_heavy_dependency_like_tensorflow import something_amazing
+
+    class MyWidget(QWidget):
+        def do_something_amazing(self):
+            return something_amazing()
+
+    @napari_hook_specification
+    def napari_experimental_provide_dock_widget():
+        return MyWidget
+    ```
+
+    This can deterioate the end-user experience, and make napari feel slugish. Best practice is to delay heavy imports until right before they are used.  The following slight modification will help napari load much faster:
+
+    ```py
+    # mypackage/napari_plugin.py
+    from napari_plugin_engine import napari_hook_specification
+    from qtpy.QtWidgets import QWidget
+
+    class MyWidget(QWidget):
+        def do_something_amazing(self):
+            # import has been moved here, will happen only after the user
+            # has opened and used this widget.
+            from my_heavy_dependency_like_tensorflow import something_amazing
+
+            return something_amazing()
+    ```
+
+    (again, the second gen napari plugin engine will help improve this situation,
+    but it's still a good idea!)
+
+- **Don't leave resources open**
+
+    It's always good practice to clean up resources like open file handles and
+    databases.  As a napari plugin it's particularly important to do this (and
+    especially for Windows users).  If someone tries to use the built-in plugin manager
+    to *uninstall* your plugin, open file handles and resources may cause the
+    process to fail or even leave your plugin in an "installed-but-unuseable"
+    state.
+    
+    Don't do this:
+
+    ```py
+    # my_plugin/module.py
+    import json
+
+    data_file = open("some_data_in_my_plugin.json")
```

Comment:
This seems like something we could also catch via a sanitization tool :)

Reply:
really?  this one I'm less clear on.  how do we reliably detect whether the closed it manually with `.close()`... name changes would be hard to detect there

Reply:
Just looking for open( could raise a warning. But yes you are right it could tricky